src/Storages/TableZnodeInfo.cpp

```diff
@@ -0,0 +1,135 @@
+    if (info.expanded_database || info.expanded_table)
+        res.renaming_restrictions = RenamingRestrictions::DO_NOT_ALLOW;
+    else if (info.expanded_uuid)
+        res.renaming_restrictions = RenamingRestrictions::ALLOW_PRESERVING_UUID;
+
+    res.zookeeper_name = zkutil::extractZooKeeperName(res.full_path);
+    res.path = zkutil::extractZooKeeperPath(res.full_path, /* check_starts_with_slash */ mode <= LoadingStrictnessLevel::CREATE, getLogger(table_id.getNameForLogs()));
+    res.path_prefix_for_drop = res.path;
+
+    if (expanded_uuid_in_path)
+    {
+        /// When dropping table with znode path "/foo/{uuid}/bar/baz", delete not only
+        /// "/foo/{uuid}/bar/baz" but also "/foo/{uuid}/bar" and "/foo/{uuid}" if they became empty.
+        ///
+        /// (We find the uuid substring by searching instead of keeping track of it when expanding
+        ///  the macro. So in principle we may find a uuid substring that wasn't expanded from a
+        ///  macro. This should be ok because we're searching for the *last* occurrence, so we'll get
+        ///  a prefix at least as long as the correct one, so we won't delete znodes outside the
+        ///  {uuid} path component. This sounds sketchy, but propagating string indices through macro
+        ///  expansion passes is sketchy too (error-prone and more complex), and on balance this seems
+        ///  better.)
+        String uuid_str = toString(table_id.uuid);
+        size_t i = res.path.rfind(uuid_str);
+        if (i == String::npos)
+            /// Possible if the macro is in the "<auxiliary_zookeeper_name>:/" prefix, but we probably
+            /// don't want to allow that.
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Can't find uuid in zookeeper path after expanding {{uuid}} macro: {} (uuid {})", res.path, uuid_str);
+        i += uuid_str.size();
+        /// In case the path is "/foo/pika{uuid}chu/bar" (or "/foo/{uuid}{replica}/bar").
+        while (i < res.path.size() && res.path[i] != '/')
+            i += 1;
+        res.path_prefix_for_drop = res.path.substr(0, i);
+    }
+
+    return res;
+}
+
+void TableZnodeInfo::dropAncestorZnodesIfNeeded(const zkutil::ZooKeeperPtr & zookeeper) const
+{
+    chassert(path.starts_with(path_prefix_for_drop));
+    if (path_prefix_for_drop.empty() || path_prefix_for_drop.size() == path.size())
+        return;
+    chassert(path[path_prefix_for_drop.size()] == '/');
+
+    String path_to_remove = path;
+    while (path_to_remove.size() > path_prefix_for_drop.size())
+    {
+        size_t i = path_to_remove.find_last_of('/');
+        chassert(i != String::npos && i >= path_prefix_for_drop.size());
+        path_to_remove = path_to_remove.substr(0, i);
+
+        Coordination::Error rc = zookeeper->tryRemove(path_to_remove);
+        if (rc != Coordination::Error::ZOK)
+            /// Znode not empty or already removed by someone else.
```

Comment:
Wdyt about checking the return code and printing a debug message if it's neither `ZNOTEMPTY` nor `ZNONODE`.

Reply:
`tryRemove` throws in that case:
```
    /// Doesn't throw in the following cases:
    /// * The node doesn't exist
    /// * Versions don't match
    /// * The node has children.
```
("Versions don't match" should never happen if version = -1, but client doesn't check that; maybe it should.)